pallets/capacity/src/lib.rs

```diff
@@ -644,18 +949,292 @@ impl<T: Config> Pallet<T> {
+			T::WeightInfo::start_new_epoch_if_needed()
 		} else {
 			// 1 for get_current_epoch_info, 1 for get_epoch_length
-			T::DbWeight::get().reads(2u64).saturating_add(RocksDbWeight::get().writes(1))
+			T::DbWeight::get().reads(2).saturating_add(RocksDbWeight::get().writes(1))
+		}
+	}
+
+	fn start_new_reward_era_if_needed(current_block: BlockNumberFor<T>) -> Weight {
+		let current_era_info: RewardEraInfo<RewardEra, BlockNumberFor<T>> = Self::get_current_era(); // 1r
+
+		if current_block.saturating_sub(current_era_info.started_at) >= T::EraLength::get().into() {
+			// 1r
+			let new_era_info = RewardEraInfo {
+				era_index: current_era_info.era_index.saturating_add(One::one()),
+				started_at: current_block,
+			};
+			CurrentEraInfo::<T>::set(new_era_info); // 1w
+
+			// carry over the current reward pool total
+			let current_reward_pool_total: BalanceOf<T> = CurrentEraProviderBoostTotal::<T>::get(); // 1
+			Self::update_provider_boost_reward_pool(
+				current_era_info.era_index,
+				current_reward_pool_total,
+			);
+			T::WeightInfo::start_new_reward_era_if_needed()
+		} else {
+			T::DbWeight::get().reads(1)
+		}
+	}
+
+	/// attempts to increment number of retargets this RewardEra
+	/// Returns:
+	///     Error::MaxRetargetsExceeded if they try to retarget too many times in one era.
+	fn update_retarget_record(staker: &T::AccountId) -> Result<(), DispatchError> {
+		let current_era: RewardEra = Self::get_current_era().era_index;
+		let mut retargets = Self::get_retargets_for(staker).unwrap_or_default();
+		ensure!(retargets.update(current_era).is_some(), Error::<T>::MaxRetargetsExceeded);
+		Retargets::<T>::set(staker, Some(retargets));
+		Ok(())
+	}
+
+	/// Performs the work of withdrawing the requested amount from the old staker-provider target details, and
+	/// from the Provider's capacity details, and depositing it into the new staker-provider target details.
+	pub(crate) fn do_retarget(
+		staker: &T::AccountId,
+		from_msa: &MessageSourceId,
+		to_msa: &MessageSourceId,
+		amount: &BalanceOf<T>,
+	) -> Result<(), DispatchError> {
+		let staking_type = Self::get_staking_account_for(staker).unwrap_or_default().staking_type;
+		let capacity_withdrawn = Self::reduce_capacity(staker, *from_msa, *amount, staking_type)?;
+
+		let mut to_msa_target = Self::get_target_for(staker, to_msa).unwrap_or_default();
+
+		to_msa_target
+			.deposit(*amount, capacity_withdrawn)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		let mut capacity_details = Self::get_capacity_for(to_msa).unwrap_or_default();
+		capacity_details
+			.deposit(amount, &capacity_withdrawn)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		Self::set_target_details_for(staker, *to_msa, to_msa_target);
+		Self::set_capacity_for(*to_msa, capacity_details);
+		Ok(())
+	}
+
+	/// updates or inserts a new boost history record for current_era.   Pass 'add' = true for an increase (provider_boost),
+	/// pass 'false' for a decrease (unstake)
+	pub(crate) fn upsert_boost_history(
```

Comment:
I think we should avoid writing upsert functions whenever possible. While they can be useful in certain situations, I think we can take a different approach here. We’ve encountered multiple bugs in our code due to this style.

Also, when calling an upsert function, it typically creates or updates fields, but in this case, it’s also removing data, which was something I didn’t expect. This could lead to unexpected behavior and further complicates the logic.

Reply:
@wilwade and I reviewed this one; this history is a ring history, I think that certainly a rename to indicate the record could be deleted is a good idea.  However it would be good to know what bugs have been seen due to this pattern so I/we can know how best to change it.

Reply:
all of our extrinsic that have upsert have had bugs.

Reply:
Pushed to an update. See #2207